### ISSUE TITLE
Support for SVG2 auto-flowed text

### ISSUE THREAD

**Issue by ernierasta** [lifelike/countersheetsextension#67]
Hi Pelle!
Currently, extension detects textflow and if found - formats it (replaces \n to actual newline). Problem is, that newer Inkscape (in my case 1.0) has as default set:
**Edit > Preferences > Tools > Text > Use SVG2 auto-flowed text**
Which means, that if you drag text area - it will still be text type, not textflow and it took me some time to realize, why my newlines are not working.

Do you think it would be possible to format this new auto-flowed text? Or should just be documented?

**Comment by lifelike** [comment]
I was not aware of this new feature and would have to investigate. Of course it would be bad to see support for multi-line text broken. Hopefully the new markup is reasonably straightforward to use without (as in some other cases...) having to replicate a tonne of logic that Inkscape already does (because of the lack of proper plugin API).

**Comment by ernierasta** [comment]
Looks like that:
```xml
    <text
       xml:space="preserve"
       id="text833"
       style="font-size:4.23333px;line-height:1.25;font-family:'I Love Christmas';-inkscape-font-specification:'I Love Christmas';letter-spacing:0px;word-spacing:0px;white-space:pre;shape-inside:url(#rect835);"><tspan
         x="46.515625"
         y="52.087409"><tspan>first line
</tspan></tspan><tspan
         x="46.515625"
         y="57.379073"><tspan>second line
</tspan></tspan><tspan
         x="46.515625"
         y="62.670737"><tspan>third line</tspan></tspan></text>

```
**#rect835** is defined earlier.
But how would anyone calculate those x,y ... this is black magic for me :)

**Comment by lifelike** [comment]
Y can probably be calculated from text size and/or line-height. X looks trivial as long as text is left-aligned. If text is centered or right-aligned I guess it will be much more difficult (or for left-to-right languages...)? But even worse is to calculate where to break the text. As long as there is manual newlines marked up in the text it will be possible, but otherwise I have no clue what the extension can do to calculate the width of rendered text to properly line-break it. That is the kind of thing Inkscape does internally but does not offer an API for as far as I know.

I wonder if there is a good workaround to make sure that templates created in Inkscape 1.0 or later use the old type of textflow areas, instead of this text with offset tspan elements?

**Comment by ernierasta** [comment]
Today will be a bit hectic for me(in a good way - Christmas!). 😃 But yes, there is, you can just disable this feature in settings. Maybe we should ask Inkscape guys about that, I will make an issue there and link it here.

**Comment by ernierasta** [comment]
It seem that situation is better then it looks initially. I had just tested this:
```xml
    <text
       xml:space="preserve"
       id="text862"
       style="font-style:normal;font-variant:normal;font-weight:normal;font-stretch:normal;font-size:4.23333px;line-height:1.25;font-family:'DejaVu Sans Mono';-inkscape-font-specification:'DejaVu Sans Mono';letter-spacing:0px;word-spacing:0px;white-space:pre;shape-inside:url(#rect864);">
       <tspan>Hello world! How are you?</tspan>
    </text>
```
And text is auto-flowed as it should, in my case, rectangle spans it on two lines.

And now, bonus, which may save the day:
```xml
xml:space="preserve"
```
means probably (it should be **white-space** with value **pre** according to spec. draft ... but it is a draft, maybe it changed) that white spaces must be preserved, so:
```xml
<tspan>Hello 
       world! 
       How are
       you?</tspan>
```
works as expected! New lines are preserved. You just need rectangle large enough.

Here is official SVG2 draft:
https://www.w3.org/TR/SVG2/text.html#TextElement

**Comment by ernierasta** [comment]
BTW: tested also in Firefox, still renders as expected, so looks promising.

So I will not create any issue, as it seems, that it may be even simpler then old textflow.

**Comment by lifelike** [comment]
When I tested this in Inkscape 1.0.1, default preferences, it created a text element with a single tspan inside, and the text had linebreaks (like expected from preserved white-space) so it looks like the default would be useful. Not sure what you did to get multiple tspan with individual x,y coordinates? That looks like the only variation that would be difficult to support.

**Comment by ernierasta** [comment]
Did you check saved file? XML Editor in Inkscape shows it differently as you say, but I have the same version of Inkscape as you and it looks always like that. But this is not important, important is, that this simple form written to file works as expected.

**Comment by lifelike** [comment]
Sounds good. So is there anything to do here, or can this be closed?

I have not seen any issues with texts, but most of my templates are much older than 1.0.

**Comment by ernierasta** [comment]
I think there is, I have not looked at code, but I will later. This issue exists, because my bold text was misaligned and new lines were ignored (they shows as **firstword\nsecondword**), so I think currently there are changes to be made. I will specify it later, or propose PR if you want. ;-)

**Comment by ernierasta** [comment]
Ok, I think I solved the problem. It will be quite trivial change at the end. After different tests I think, that inserting newline into string is best solution. It works good (new lines are where they should, additionally auto-wrapping works, bold\italic alignment is fixed), generated PDF files are as the should. The only thing I wasn't happy about was Inkscape throwing an error while exporting to PDF:
`Invalid glyph found, continuing...`
on every encountered newline character.

I wanted to filter out this, feeling guilty, that this not optimal ... then I found *print_filtered_stderr* in code and said to myself: "Well I am not only one!". :-)

And when you save file, auto-flowed text is changed to `<tspan>`s with exact x,y (from what I read it done like that because of backward compatability). 

Now I just need to start fresh and create proper PR and look deeper into all code paths, as it is possible that some processing is no longer necessary.
